progress outputs

Hello FluCoMites,

It would be nice to have progress outputs from stdout perhaps issued with a verbose flag or something ala Max environment objects. I took a look at implementing myself - alas I lack the C++ skills to really get my head around it but perhaps putting it on your radar might make it find its way into a future release. With some guidance + clarity on how to implement it could be something I take on myself.

This is what I get back:

```clang
In file included from /Users/james/dev/flucoma/flucoma-cli/src/fluid-ampgate/fluid-ampgate.cpp:12:
/Users/james/dev/flucoma/flucoma-cli/include/FluidCLIWrapper.hpp:393:42: error: too many arguments to function call, expected 0, have 1
    auto         result = client.process(context);
                          ~~~~~~~~~~~~~~ ^~~~~~~
/Users/james/dev/flucoma/flucoma-cli/src/fluid-ampgate/fluid-ampgate.cpp:18:48: note: in instantiation of member function
      'fluid::client::CLIWrapper<NRTThreadedAmpGateClient>::run' requested here
  return CLIWrapper<NRTThreadedAmpGateClient>::run(argc, argv);
                                               ^
/Users/james/dev/flucoma/flucoma-cli/build/_deps/flucoma-core-src/include/clients/rt/../common/FluidNRTClientWrapper.hpp:522:3: note: 'process' declared here
  Result process()
  ^
1 error generated.
make[2]: *** [src/fluid-ampgate/CMakeFiles/fluid-ampgate.dir/fluid-ampgate.cpp.o] Error 1
make[1]: *** [src/fluid-ampgate/CMakeFiles/fluid-ampgate.dir/all] Error 2
make: *** [all] Error 2
```

Uh oh! Is this an easy one you've come across? I did `brew update && brew upgrade` not too long ago so perhaps it updated one of the underlying libs required for compilation?

-- The CXX compiler identification is AppleClang 11.0.0.11000033
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - failed
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ - broken
CMake Error at /usr/local/Cellar/cmake/3.18.2/share/cmake/Modules/CMakeTestCXXCompiler.cmake:59 (message):
  The C++ compiler

    "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /Users/james/dev/flucoma/flucoma-cli/build/CMakeFiles/CMakeTmp

    Run Build Command(s):/usr/bin/make cmTC_e034f/fast && /Applications/Xcode.app/Contents/Developer/usr/bin/make  -f CMakeFiles/cmTC_e034f.dir/build.make CMakeFiles/cmTC_e034f.dir/build
    Building CXX object CMakeFiles/cmTC_e034f.dir/testCXXCompiler.cxx.o
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++   -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -mmacosx-version-min=10.8 -o CMakeFiles/cmTC_e034f.dir/testCXXCompiler.cxx.o -c /Users/james/dev/flucoma/flucoma-cli/build/CMakeFiles/CMakeTmp/testCXXCompiler.cxx
    warning: include path for stdlibc++ headers not found; pass '-stdlib=libc++' on the command line to use the libc++ standard library instead [-Wstdlibcxx-not-found]
    1 warning generated.
    Linking CXX executable cmTC_e034f
    /usr/local/Cellar/cmake/3.18.2/bin/cmake -E cmake_link_script CMakeFiles/cmTC_e034f.dir/link.txt --verbose=1
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -mmacosx-version-min=10.8 -Wl,-search_paths_first -Wl,-headerpad_max_install_names CMakeFiles/cmTC_e034f.dir/testCXXCompiler.cxx.o -o cmTC_e034f
    clang: warning: libstdc++ is deprecated; move to libc++ with a minimum deployment target of OS X 10.9 [-Wdeprecated]
    ld: library not found for -lstdc++
    clang: error: linker command failed with exit code 1 (use -v to see invocation)
    make[1]: *** [cmTC_e034f] Error 1
    make: *** [cmTC_e034f/fast] Error 2


  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:25 (project)


-- Configuring incomplete, errors occurred!
See also "/Users/james/dev/flucoma/flucoma-cli/build/CMakeFiles/CMakeOutput.log".
See also "/Users/james/dev/flucoma/flucoma-cli/build/CMakeFiles/CMakeError.log".

--- a/include/FluidCLIWrapper.hpp
+++ b/include/FluidCLIWrapper.hpp
@@ -46,7 +46,6 @@ public:
     if (file.isOpen())
     {
       resize(file.getFrames(), file.getChannels(), file.getSamplingRate());
-
       file.seek();
       file.readInterleaved(mData.data(),file.getFrames());
     }
@@ -88,11 +87,12 @@ public:
       uint16_t chans = static_cast<uint16_t>(
           std::min(asSigned(std::numeric_limits<uint16_t>::max()), numChans()));
 
-      HISSTools::OAudioFile file(mPath, fileType, depthType, chans,
+        HISSTools::OAudioFile file(mPath, fileType, depthType, chans,
                                  mSamplingRate);
 
       if (file.isOpen())
       {
+        std::cout << "Writing " << mPath << '\n';
         file.seek();
         file.writeInterleaved(mData.data(), static_cast<uint32_t>(numFrames()));
         
@@ -423,29 +423,28 @@ public:
     if(!readSuccess) return -1; 
 
     // Create client after all parameters are set
-        // Create client after all parameters are set
+
     ClientType   client(params);
     Result result;
 
     client.enqueue(params);
     result = client.process();
     
-    double progress = 0.0; // Variable to store progress
+    double progress = 0.0;
 
     while(result.ok())
     {
         ProcessState state = client.checkProgress(result);
       
         if (state == ProcessState::kDone || state == ProcessState::kDoneStillProcessing) {
-          std::cout << '\n';
+          std::cout << "100%\n";
           break;
         }
         if (state != ProcessState::kDone) {
           double newProgress = client.progress();
           if (newProgress - progress >=0.01)
           {
-            std::cout << newProgress << '\r' << std::flush;
-            
+            std::cout << std::setw(3)  << static_cast<int>(100 * newProgress) << "%\r" << std::flush;
             progress = newProgress;
           }
           using namespace std::chrono_literals;
@@ -457,11 +456,13 @@ public:
     if (!result.ok())
     {
       // Output error
+      
       std::cerr << result.message() << "\n";
     }
     else
     {
       // Write files
+      
       bool allowCSV = true;
       bool fileWriteResult = true;
       params.template forEachParamType<BufferT, WriteFiles>(allowCSV, fileWriteResult);
@@ -474,5 +475,3 @@ public:
 
 } // namespace client
 } // namespace fluid
-
-

--- a/include/FluidCLIWrapper.hpp
+++ b/include/FluidCLIWrapper.hpp
@@ -390,23 +390,32 @@ public:
     params.constrainParameterValues();
 
     // Create client after all parameters are set
-    FluidContext context;
+        // Create client after all parameters are set
     ClientType   client(params);
     Result result;
-    using namespace std::chrono_literals;
 
-    client.process();
-    ProcessState state = client.checkProgress(result);
+    client.enqueue(params);
+    result = client.process();
+    
     double progress = 0.0; // Variable to store progress
 
-    while(true) 
+    while(result.ok())
     {
+        ProcessState state = client.checkProgress(result);
+      
         if (state == ProcessState::kDone || state == ProcessState::kDoneStillProcessing) {
+          std::cout << '\n';
           break;
         }
         if (state != ProcessState::kDone) {
-          if (client.progress() - progress >=1) std::cout << client.progress() << "\n";
-          progress = client.progress();
+          double newProgress = client.progress();
+          if (newProgress - progress >=0.01)
+          {
+            std::cout << newProgress << '\r' << std::flush;
+            
+            progress = newProgress;
+          }
+          using namespace std::chrono_literals;
           std::this_thread::sleep_for(20ms); 
           continue; 
         }
@@ -430,3 +439,5 @@ public:
 
 } // namespace client
 } // namespace fluid
+
+

--- a/include/FluidCLIWrapper.hpp
+++ b/include/FluidCLIWrapper.hpp
@@ -18,10 +18,12 @@ under the European Union’s Horizon 2020 research and innovation programme
 #include <clients/common/ParameterTypes.hpp>
 #include <FluidVersion.hpp>
 #include <cctype>
+#include <chrono>
 #include <iomanip>
 #include <iostream>
 #include <regex>
 #include <string>
+#include <thread>
 #include <utility>
 #include <vector>
 

--- a/include/FluidCLIWrapper.hpp
+++ b/include/FluidCLIWrapper.hpp
@@ -392,18 +392,34 @@ public:
     // Create client after all parameters are set
     FluidContext context;
     ClientType   client(params);
-    auto         result = client.process(context);
+    Result result;
+    using namespace std::chrono_literals;
+
+    client.process();
+    ProcessState state = client.checkProgress(result);
+    double progress = 0.0; // Variable to store progress
+
+    while(true) 
+    {
+        if (state == ProcessState::kDone || state == ProcessState::kDoneStillProcessing) {
+          break;
+        }
+        if (state != ProcessState::kDone) {
+          if (client.progress() - progress >=1) std::cout << client.progress() << "\n";
+          progress = client.progress();
+          std::this_thread::sleep_for(20ms); 
+          continue; 
+        }
+    }
 
     if (!result.ok())
     {
       // Output error
-
       std::cerr << result.message() << "\n";
     }
     else
     {
       // Write files
-
       bool allowCSV = true;
       params.template forEachParamType<BufferT, WriteFiles>(allowCSV);
     }

--- a/src/fluid-ampgate/fluid-ampgate.cpp
+++ b/src/fluid-ampgate/fluid-ampgate.cpp
@@ -15,5 +15,5 @@ under the European Union’s Horizon 2020 research and innovation programme
 int main(int argc, const char* argv[])
 {
   using namespace fluid::client;
-  return CLIWrapper<NRTAmpGateClient>::run(argc, argv);
+  return CLIWrapper<NRTThreadedAmpGateClient>::run(argc, argv);
 }

--- a/src/fluid-ampslice/fluid-ampslice.cpp
+++ b/src/fluid-ampslice/fluid-ampslice.cpp
@@ -15,5 +15,5 @@ under the European Union’s Horizon 2020 research and innovation programme
 int main(int argc, const char* argv[])
 {
   using namespace fluid::client;
-  return CLIWrapper<NRTAmpSliceClient>::run(argc, argv);
+  return CLIWrapper<NRTThreadedAmpSliceClient>::run(argc, argv);
 }

--- a/src/fluid-hpss/fluid-hpss.cpp
+++ b/src/fluid-hpss/fluid-hpss.cpp
@@ -15,5 +15,5 @@ under the European Union’s Horizon 2020 research and innovation programme
 int main(int argc, const char* argv[])
 {
   using namespace fluid::client;
-  return CLIWrapper<NRTHPSSClient>::run(argc, argv);
+  return CLIWrapper<NRTThreadedHPSSClient>::run(argc, argv);
 }

--- a/src/fluid-loudness/fluid-loudness.cpp
+++ b/src/fluid-loudness/fluid-loudness.cpp
@@ -15,5 +15,5 @@ under the European Union’s Horizon 2020 research and innovation programme
 int main(int argc, const char* argv[])
 {
   using namespace fluid::client;
-  return CLIWrapper<NRTLoudnessClient>::run(argc, argv);
+  return CLIWrapper<NRTThreadedLoudnessClient>::run(argc, argv);
 }

--- a/src/fluid-melbands/fluid-melbands.cpp
+++ b/src/fluid-melbands/fluid-melbands.cpp
@@ -15,5 +15,5 @@ under the European Union’s Horizon 2020 research and innovation programme
 int main(int argc, const char* argv[])
 {
   using namespace fluid::client;
-  return CLIWrapper<NRTMelBandsClient>::run(argc, argv);
+  return CLIWrapper<NRTThreadedMelBandsClient>::run(argc, argv);
 }

--- a/src/fluid-mfcc/fluid-mfcc.cpp
+++ b/src/fluid-mfcc/fluid-mfcc.cpp
@@ -15,5 +15,5 @@ under the European Union’s Horizon 2020 research and innovation programme
 int main(int argc, const char* argv[])
 {
   using namespace fluid::client;
-  return CLIWrapper<NRTMFCCClient>::run(argc, argv);
+  return CLIWrapper<NRTThreadedMFCCClient>::run(argc, argv);
 }

--- a/src/fluid-nmf/fluid-nmf.cpp
+++ b/src/fluid-nmf/fluid-nmf.cpp
@@ -15,5 +15,5 @@ under the European Union’s Horizon 2020 research and innovation programme
 int main(int argc, const char* argv[])
 {
   using namespace fluid::client;
-  return CLIWrapper<NMFClient>::run(argc, argv);
+  return CLIWrapper<NRTThreadedNMFClient>::run(argc, argv);
 }

--- a/src/fluid-noveltyslice/fluid-noveltyslice.cpp
+++ b/src/fluid-noveltyslice/fluid-noveltyslice.cpp
@@ -15,5 +15,5 @@ under the European Union’s Horizon 2020 research and innovation programme
 int main(int argc, const char* argv[])
 {
   using namespace fluid::client;
-  return CLIWrapper<NRTNoveltySliceClient>::run(argc, argv);
+  return CLIWrapper<NRTThreadingNoveltySliceClient>::run(argc, argv);
 }

--- a/src/fluid-onsetslice/fluid-onsetslice.cpp
+++ b/src/fluid-onsetslice/fluid-onsetslice.cpp
@@ -15,5 +15,5 @@ under the European Union’s Horizon 2020 research and innovation programme
 int main(int argc, const char* argv[])
 {
   using namespace fluid::client;
-  return CLIWrapper<NRTOnsetSliceClient>::run(argc, argv);
+  return CLIWrapper<NRTThreadingOnsetSliceClient>::run(argc, argv);
 }

--- a/src/fluid-pitch/fluid-pitch.cpp
+++ b/src/fluid-pitch/fluid-pitch.cpp
@@ -15,5 +15,5 @@ under the European Union’s Horizon 2020 research and innovation programme
 int main(int argc, const char* argv[])
 {
   using namespace fluid::client;
-  return CLIWrapper<NRTPitchClient>::run(argc, argv);
+  return CLIWrapper<NRTThreadedPitchClient>::run(argc, argv);
 }

--- a/src/fluid-sines/fluid-sines.cpp
+++ b/src/fluid-sines/fluid-sines.cpp
@@ -15,5 +15,5 @@ under the European Union’s Horizon 2020 research and innovation programme
 int main(int argc, const char* argv[])
 {
   using namespace fluid::client;
-  return CLIWrapper<NRTSinesClient>::run(argc, argv);
+  return CLIWrapper<NRTThreadedSinesClient>::run(argc, argv);
 }

--- a/src/fluid-spectralshape/fluid-spectralshape.cpp
+++ b/src/fluid-spectralshape/fluid-spectralshape.cpp
@@ -15,5 +15,5 @@ under the European Union’s Horizon 2020 research and innovation programme
 int main(int argc, const char* argv[])
 {
   using namespace fluid::client;
-  return CLIWrapper<NRTSpectralShapeClient>::run(argc, argv);
+  return CLIWrapper<NRTThreadedSpectralShapeClient>::run(argc, argv);
 }

--- a/src/fluid-stats/fluid-stats.cpp
+++ b/src/fluid-stats/fluid-stats.cpp
@@ -15,5 +15,5 @@ under the European Union’s Horizon 2020 research and innovation programme
 int main(int argc, const char* argv[])
 {
   using namespace fluid::client;
-  return CLIWrapper<BufStatsClient>::run(argc, argv);
+  return CLIWrapper<NRTThreadedBufStatsClient>::run(argc, argv);
 }

--- a/src/fluid-transients/fluid-transients.cpp
+++ b/src/fluid-transients/fluid-transients.cpp
@@ -15,5 +15,5 @@ under the European Union’s Horizon 2020 research and innovation programme
 int main(int argc, const char* argv[])
 {
   using namespace fluid::client;
-  return CLIWrapper<NRTTransientsClient>::run(argc, argv);
+  return CLIWrapper<NRTThreadedTransientsClient>::run(argc, argv);
 }

--- a/src/fluid-transientslice/fluid-transientslice.cpp
+++ b/src/fluid-transientslice/fluid-transientslice.cpp
@@ -15,5 +15,7 @@ under the European Union’s Horizon 2020 research and innovation programme
 int main(int argc, const char* argv[])
 {
   using namespace fluid::client;
-  return CLIWrapper<NRTTransientSliceClient>::run(argc, argv);
+  // return CLIWrapper<NRTTransientSliceClient>::run(argc, argv);
+  return CLIWrapper<NRTThreadedTransientSliceClient>::run(argc, argv);
+
 }

